### ISSUE TITLE
OpenSCAD parser error in the *middle* of a comment when last line accidentally uncommented?

### ISSUE THREAD

**Issue by Jimw338** [openscad/openscad#3939]
I have the following segment of comment in my code file (the last line is intentionally missing the slash characters - that the error):

```
// * Cube Back Text from Méi guī
// 一颗 心 怕 摔倒 打破
//"yī kē xīn  pà shuāi dǎo  dǎ pò"
// "It's the heart afraid of breaking"
// 不敢 起舞 欢 歌
// "bù gǎn  qǐwǔ huāngē"
// "that never learns to dance."

不敢起舞欢歌				(dance)
```

But the red error dot appears on the second character of the second line (second slash of the line "// * Cube Back Text from Méi guī".

Stranger, if I remove various quotes in the comment, the character/line that is hilited changes. (and sometimes - don't know why - it actually ends up on the right line)

The parser seems (I'm guessing) to some kind of recursive structure that is continuing to parse the quotes *inside the comment itself*, and ultimately fails and ends up placing the error dot where it bailed out, which is not where the actual error was.

Shouldn't the parser be set up to completely ignore *anything* past the second slash on a commented line (until the end-of-line), or have a "special" function to parse comments, so that the error tag shows up on the line where the error actually is?

As it is, I futzed around about 10 minutes wondering if OpenSCAD just didn't like files longer than 575 lines.

**Comment by jordanbrown0** [comment]
There should be an associated error message; what line did that message report?

When I tried it (with just the snippet supplied) it correctly reported that the error was on line 9, but then put the red dot on line 6.  I assume that the red dot is the result of some post-processing that tries to figure out from the error report what line in the editor to highlight.

As I incrementally delete the non-ASCII characters, the highlight moves towards the correct position.  I suspect that something is counting a position in characters, and then using that value as a position in bytes.  Each of the non-ASCII characters is several bytes, so the position would be off.

Here's a trivial example:
```
// 🦃🦃abcdefghi
x;
```
The error is on line 2, but the highlight is on line 1, on the "f".  The turkeys are each four bytes, so you would expect it to miscount by 6.  Absent the turkeys the highlight is on the semicolon, and (counting newline as one byte) 6 bytes earlier is the "f".

**Comment by t-paul** [comment]
This has nothing to do with the `comment_parser`, this is part of the customizer. The issue is that `flex` does not support unicode so there's some trickery trying to work around that issue. So the place that likely need to be fixed is `lexer.l`.

**Comment by t-paul** [comment]
Right, there was a similar issue with the find indicators which made https://github.com/openscad/openscad/blob/master/src/ScintillaEditor.cc#L655-L675 look a bit complicated. So basically I think we need to find a solution that correctly syncs understanding of positions between parser and editor.

**Comment by ChrisCoxArt** [comment]
The error_pos coming in appears to be in glyphs, because it is short by exactly the number of 2 byte chars before it, including comments.
lineIndexFromPosition appears to want bytes.
For example
```
// a€
abc•=5;
```
The error_pos is 9, which is the location of 'b' not the dot. And 'b' gets highlighted as the error.
In a longer example, the line number would also be off because of undercounted bytes.

**Comment by ChrisCoxArt** [comment]
It appears that several unicode specific adjustments to parser_error_pos in lexer.l are unnecessary and can be removed to get correct positions (and line numbers).  
There are still some problems with error highlighting of 4 byte unicode characters: the editor doesn't always want to remove the red highlight on them after fixing the string/comment.  But so far I can only reproduce it sometimes, with some Chinese characters. It seems to have to do with preceeding 4 byte characters.
Eh, let's get this basic bug fixed first.

**Comment by ChrisCoxArt** [comment]
pull request in progress.

**Comment by t-paul** [comment]
Fixed via #4049.

![grafik](https://user-images.githubusercontent.com/1330241/149629460-d24cce89-ca12-41bb-b70d-ed320d91458a.png)
